### ISSUE TITLE
ConcurrentModificationException in EclipseLink ObjectExpression

### ISSUE THREAD

**Issue by alexbudarov** [cuba-platform/cuba#1818]
### Environment
- Platform version: 6.8.14
- EclipseLink: 2.6.2.cuba25

### Description of the bug or enhancement
A rare exception: occured three times today on one of servers. Have not seen it before.
<details>
 <summary>Click to expand</summary>
 
```js
2019-01-25 09:23:21,236 ERROR [http-nio-8080-exec-70/app-core/jenny] [com.haulmont.cuba.core.sys.ServiceInterceptor] Exception: 
javax.persistence.PersistenceException: Exception [EclipseLink-6168] (Eclipse Persistence Services - 2.6.2.cuba25): org.eclipse.persistence.exceptions.QueryException
Exception Description: Query failed to prepare, unexpected error occurred: [java.util.ConcurrentModificationException].
Internal Exception: java.util.ConcurrentModificationException
Query: ReadObjectQuery(referenceClass=Xxxx sql="SELECT ID, DTYPE, .... FROM XXXX_JOB WHERE ((ID = ?) AND ((DELETE_TS IS NULL) AND (DTYPE = ?)))")
FetchGroup(){...}
        at org.eclipse.persistence.internal.jpa.QueryImpl.getSingleResult(QueryImpl.java:551)
        at org.eclipse.persistence.internal.jpa.EJBQueryImpl.getSingleResult(EJBQueryImpl.java:402)
        at com.haulmont.cuba.core.sys.QueryImpl.getResultFromCache(QueryImpl.java:647)
        at com.haulmont.cuba.core.sys.QueryImpl.getSingleResult(QueryImpl.java:361)
        at com.haulmont.cuba.core.app.RdbmsStore.executeQuery(RdbmsStore.java:692)
        at com.haulmont.cuba.core.app.RdbmsStore.load(RdbmsStore.java:136)
        at com.haulmont.cuba.core.app.DataManagerBean.load(DataManagerBean.java:68)
...
        at sun.reflect.GeneratedMethodAccessor2460.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:333)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:190)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
        at org.springframework.aop.aspectj.MethodInvocationProceedingJoinPoint.proceed(MethodInvocationProceedingJoinPoint.java:85)
        at com.haulmont.cuba.core.sys.ServiceInterceptor.aroundInvoke(ServiceInterceptor.java:117)
        at sun.reflect.GeneratedMethodAccessor975.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.aop.aspectj.AbstractAspectJAdvice.invokeAdviceMethodWithGivenArgs(AbstractAspectJAdvice.java:629)
        at org.springframework.aop.aspectj.AbstractAspectJAdvice.invokeAdviceMethod(AbstractAspectJAdvice.java:618)
        at org.springframework.aop.aspectj.AspectJAroundAdvice.invoke(AspectJAroundAdvice.java:70)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:168)
        at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:92)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
        at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:213)
        at com.sun.proxy.$Proxy337.getDriverDetails(Unknown Source)
...
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
        at java.lang.Thread.run(Thread.java:748)
Caused by: org.eclipse.persistence.exceptions.QueryException: 
Exception Description: Query failed to prepare, unexpected error occurred: [java.util.ConcurrentModificationException].
Internal Exception: java.util.ConcurrentModificationException
Query: ReadObjectQuery(referenceClass=Xxxx sql="SELECT ID, DTYPE, ... FROM XXXX_JOB WHERE ((ID = ?) AND ((DELETE_TS IS NULL) AND (DTYPE = ?)))")
FetchGroup(){...}
        at org.eclipse.persistence.exceptions.QueryException.prepareFailed(QueryException.java:1596)
        at org.eclipse.persistence.queries.DatabaseQuery.checkPrepare(DatabaseQuery.java:685)
        at org.eclipse.persistence.queries.ObjectLevelReadQuery.checkPrepare(ObjectLevelReadQuery.java:908)
        at org.eclipse.persistence.queries.DatabaseQuery.checkPrepare(DatabaseQuery.java:615)
        at org.eclipse.persistence.queries.DatabaseQuery.execute(DatabaseQuery.java:872)
        at org.eclipse.persistence.queries.ObjectLevelReadQuery.execute(ObjectLevelReadQuery.java:1131)
        at org.eclipse.persistence.queries.ReadObjectQuery.execute(ReadObjectQuery.java:451)
        at org.eclipse.persistence.queries.ObjectLevelReadQuery.executeInUnitOfWork(ObjectLevelReadQuery.java:1219)
        at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.internalExecuteQuery(UnitOfWorkImpl.java:2944)
        at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1854)
        at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1836)
        at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1801)
        at org.eclipse.persistence.internal.jpa.QueryImpl.executeReadQuery(QueryImpl.java:258)
        at org.eclipse.persistence.internal.jpa.QueryImpl.getSingleResult(QueryImpl.java:521)
        ... 80 common frames omitted
Caused by: java.util.ConcurrentModificationException: null
        at java.util.ArrayList$Itr.checkForComodification(ArrayList.java:909)
        at java.util.ArrayList$Itr.next(ArrayList.java:859)
        at org.eclipse.persistence.internal.expressions.ObjectExpression.isTreatUsed(ObjectExpression.java:271)
        at org.eclipse.persistence.queries.ObjectLevelReadQuery.prepareQuery(ObjectLevelReadQuery.java:2222)
        at org.eclipse.persistence.queries.ObjectLevelReadQuery.prepare(ObjectLevelReadQuery.java:1963)
        at org.eclipse.persistence.queries.ReadObjectQuery.prepare(ReadObjectQuery.java:763)
        at org.eclipse.persistence.queries.DatabaseQuery.checkPrepare(DatabaseQuery.java:666)
        ... 92 common frames omitted
```
</details>

Have not tried to reproduce it yet. But I suppose the case is: two parallel requests to DataManager to reload the same entity class by ID with two different views.

The code in ObjectExpression is pretty suspicious with modifying that collection.

**Comment by alexbudarov** [comment]
tried to reproduce but haven't succeeded

**Comment by alexbudarov** [comment]
posted but to eclipselink tracker: https://bugs.eclipse.org/bugs/show_bug.cgi?id=546812

**Comment by andreysubbotin** [comment]
Fixed in EclipseLink: 2.6.2.cuba26 and 2.7.3-4-cuba. Synchronization added.
